### ISSUE TITLE
[N:M Sparsity] No Gemm/MatMul eligible for sparse math

### ISSUE THREAD

**Issue by wangshangsam** [NVIDIA/TensorRT#2271]
## Description

I've been trying to enable sparse tensor core on A100 and RTX3090 for the BERT SQuAD inference. However, even when I can see in Netron that the `Gemm`'s weight satisfy the 2:4 sparsity pattern,
![Screenshot from 2022-08-23 14-35-32](https://user-images.githubusercontent.com/7806007/186240851-eba9d709-f92c-4e7e-ad49-a1710c1984f2.png)
and I set `--sparsity=force`, `trtexec` still shows that "Layers eligible for sparse math:" is empty. Example log from `trtexec` using the command provided in the [Steps To Reproduce](##Steps_To_Reproduce) section:
```bash
[08/23/2022-16:17:57] [I] [TRT] Detected 3 inputs and 2 output network tensors.                                                                                                              
[08/23/2022-16:18:06] [I] [TRT] Total Host Persistent Memory: 160                                                                                                                            
[08/23/2022-16:18:06] [I] [TRT] Total Device Persistent Memory: 0                                                                                                                            
[08/23/2022-16:18:06] [I] [TRT] Total Scratch Memory: 213995520                                                                                                                              
[08/23/2022-16:18:06] [I] [TRT] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 1334 MiB, GPU 2048 MiB                                                               
[08/23/2022-16:18:06] [I] [TRT] [BlockAssignment] Algorithm ShiftNTopDown took 0.381067ms to assign 5 blocks to 7 nodes requiring 214022144 bytes.                                           
[08/23/2022-16:18:06] [I] [TRT] Total Activation Memory: 214022144                                                                                                                           
[08/23/2022-16:18:06] [I] [TRT] (Sparsity) Layers eligible for sparse math:                                                                                                                  
[08/23/2022-16:18:06] [I] [TRT] (Sparsity) TRT inference plan picked sparse implementation for layers:                                                                                       
[08/23/2022-16:18:06] [I] [TRT] [MemUsageChange] TensorRT-managed allocation in building engine: CPU +0, GPU +2048, now: CPU 0, GPU 2048 (MiB)                                               
[08/23/2022-16:18:06] [W] [TRT] The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This 
function will always return 1.                                                                                                                                                               
[08/23/2022-16:18:07] [W] [TRT] The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This 
function will always return 1.                                                                                                                                                               
[08/23/2022-16:18:14] [I] Engine built in 2730.82 sec.                                                                                                                                       
```
Later when I profiled the inference iteration via nsight system, I could see that no `sparse_hmma_gemm` kernel was executed, indeed.

The ONNX checkpoint (attached below in the [Relevant Files](#Relevant_Files) section) I used is essentially generated from [DeepLearningExamples/PyTorch/LanguageModeling/BERT/run_squad.py](https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/run_squad.py). I modified https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/modeling.py a bit such that ONNX would convert `torch.functional.linear` to `Gemm` instead of `MatMul` (I heard that `trtexec` doesn't support N:M sparsity for `MatMul` currently).

I'm wondering how I can make `trtexec` to realize that those `Gemm`s are eligible for sparse math (even before asking `trtexec` select sparse tactic for those ops)?

## Environment

**TensorRT Version**: From the NGC PyTorch container below
**NVIDIA GPU**: NVIDIA A100-SXM4-40GB, NVIDIA GeForce RTX 3090
**NVIDIA Driver Version**: 495.46, 515.65.01
**CUDA Version**: From the NGC PyTorch container below
**CUDNN Version**: From the NGC PyTorch container below
**Operating System**: Ubuntu 20.04, Ubuntu 22.04
**Python Version (if applicable)**: From the NGC PyTorch container below
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: From the NGC PyTorch container below
**Baremetal or Container (if so, version)**: nvcr.io/nvidia/pytorch:22.07-py3

## Relevant Files

pytorch_model.onnx: https://drive.google.com/file/d/1jlrDsnxvCbfKYs9fchPD4KDgRncHGnlG/view?usp=sharing

## Steps To Reproduce

1. Download the ONNX checkpoint from the above link
2. Launch a NGC PyTorch container
3. Run
```bash
trtexec --onnx=pytorch_model.onnx \
  --saveEngine=pytorch_model.plan \
  --workspace=24576 \
  --minShapes=input_ids:1x384,segment_ids:1x384,input_mask:1x384 \
  --optShapes=input_ids:16x384,segment_ids:16x384,input_mask:16x384 \
  --maxShapes=input_ids:16x384,segment_ids:16x384,input_mask:16x384 \
  --fp16 --sparsity=force                    
```

**Comment by wangshangsam** [comment]
Moreover, it seems like when I set `--fp16 --sparsity=force`, the model outputs seem correct sometimes and incorrect some-other-times. For example:
```bash
root@shangw-MS-7D25:/workspace/bert# trtexec --onnx=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/pytorch_model.onnx --saveEn
gine=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/pytorch_model.plan --workspace=24576 --minShapes=input_ids:1x384,segment_i
ds:1x384,input_mask:1x384 --optShapes=input_ids:16x384,segment_ids:16x384,input_mask:16x384 --maxShapes=input_ids:16x384,segment_ids:16x384,input_mask:16x384 --fp16 --sparsity=enable 
...
root@shangw-MS-7D25:/workspace/bert# python run_squad_trt.py --engine=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/pytorch_m
odel.plan --output_dir=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/ --batch_size=16                
DLL 2022-08-23 20:28:00.187570 - PARAMETER infer_start : True                                                                                                                                
DLL 2022-08-23 20:28:00.187628 - PARAMETER eval_samples : 10570                               
DLL 2022-08-23 20:28:00.187646 - PARAMETER eval_features : 10833                                                                                                                             
DLL 2022-08-23 20:28:00.187658 - PARAMETER predict_batch_size : 16                                                                                                                           
DLL 2022-08-23 20:28:00.762998 - PARAMETER eval_start : True                                  
Evaluating: 100%|███████████████████████████| 678/678 [00:37<00:00, 18.10it/s]                                                                                                               
DLL 2022-08-23 20:29:12.419866 -  e2e_inference_time : 37.45573878288269 s inference_sequences_per_second : 289.2213677267178 sequences/s exact_match : 0.17975402081362346 None F1 : 6.9631350697915595 None mean latencies : 53.99744994169139  std latencies : 3.0545661742863666       
```
Note that the `F1` is extremely low. However, when I run the same set of commands again:
```bash
root@shangw-MS-7D25:/workspace/bert# trtexec --onnx=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/pytorch_model.onnx --saveEngine=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/pytorch_model.plan --workspace=24576 --minShapes=input_ids:1x384,segment_ids:1x384,input_mask:1x384 --optShapes=input_ids:16x384,segment_ids:16x384,input_mask:16x384 --maxShapes=input_ids:16x384,segment_ids:16x384,input_mask:16x384 --fp16 --sparsity=enable
...
root@shangw-MS-7D25:/workspace/bert# python run_squad_trt.py --engine=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/pytorch_model.plan --output_dir=results/SQuAD/test_finetuned_checkpoint_pruned_by_asp_then_refinetuned/1xNVIDIA_GeForce_RTX_3090/fp16/seed\=0/ --batch_size=16
DLL 2022-08-23 20:57:32.261449 - PARAMETER infer_start : True 
DLL 2022-08-23 20:57:32.261514 - PARAMETER eval_samples : 10570 
DLL 2022-08-23 20:57:32.261532 - PARAMETER eval_features : 10833 
DLL 2022-08-23 20:57:32.261543 - PARAMETER predict_batch_size : 16 
DLL 2022-08-23 20:57:32.846419 - PARAMETER eval_start : True 
Evaluating: 100%|█████████████████████████████████████████████████████████████████████| 678/678 [00:37<00:00, 18.00it/s]
DLL 2022-08-23 20:58:35.523031 -  e2e_inference_time : 37.675649881362915 s inference_sequences_per_second : 287.533195422298 sequences/s exact_match : 83.48155156102176 None F1 : 90.46869535595951 None mean latencies : 54.354972809816886  std latencies : 2.7915934076217335 
```

**Comment by zerollzeng** [comment]
@nvpohanh Any comments here? I think we support sparsity for GEMM, and I've seen other users ask about it too, but I don't have a certain answer.

**Comment by wangshangsam** [comment]
> I think we support sparsity for GEMM

I thought so too, so you can imgaine my disappointment after I spent a lot of time converting all the `MatMul`s to `Gemm`s and only to find out `trtexec` said no `Gemm` is "eligible for sparse math" :)

**Comment by zerollzeng** [comment]
1) we only support sparsity GEMM if it can be converted to a Conv(e.g. the second input must be a constant and only has 2 dimensions)
2) if the GEMM is in a transformer structure, then it(actually the whole structure) will be offloaded to Myelin(a close source DL compiler used by TRT). but unfortunately, Myelin doesn't support sparsity now.

**Comment by wangshangsam** [comment]
> if the GEMM is in a transformer structure, then it(actually the whole structure) will be offloaded to Myelin(a close source DL compiler used by TRT). but unfortunately, Myelin doesn't support sparsity now.

Thanks a lot for the clarification! I have two follow-up questions:
1. Hence, is the 2:4 sparse NeMo BERT-Large SQuAD v1.1 result reported in https://developer.nvidia.com/blog/accelerating-inference-with-sparsity-using-ampere-and-tensorrt/ only simulated in ASP, but not actually measured when the model is compiled with TRT?
2. When a transformer structure is detected, would TRT always offload the model to Myelin without any kernel profiling, even if no-Myelin might have better performance? Is there a user-facing way (e.g., flag or env vars) that we could use to temporarily disable Myelin?

**Comment by zerollzeng** [comment]
> 1. Hence, is the 2:4 sparse NeMo BERT-Large SQuAD v1.1 result reported in https://developer.nvidia.com/blog/accelerating-inference-with-sparsity-using-ampere-and-tensorrt/ only simulated in ASP, but not actually measured when the model is compiled with TRT?

I read this post and I think it only measures the accuracy of BERT SQUAD. the performance is tested using resnet-101 as shown below:
![image](https://user-images.githubusercontent.com/38289304/187054587-20b963c9-cf83-498e-93eb-5c0e3b9a75bd.png)

**Comment by zerollzeng** [comment]
> 2\. When a transformer structure is detected, would TRT always offload the model to Myelin without any kernel profiling, even if no-Myelin might have better performance? 

I think yes and myelin indeed has better performance for transformer structure(when doing this perf optimization we just leave it to myelin).

> Is there a user-facing way (e.g., flag or env vars) that we could use to temporarily disable Myelin?

Unfortunately, you can't :-(

**Comment by wzywzywzy** [comment]
Here I meet the similar problem
I try to use ASP to sparsity my bert model on squad V1.1
but the detect sparsity is empty , more information about my setting can see this link https://github.com/pytorch/TensorRT/issues/1340

![image](https://user-images.githubusercontent.com/32805624/191452639-fe804c57-4b88-4670-829c-54771fd8db9b.png)

**Comment by nvpohanh** [comment]
The ONNX path doesn't support sparsity for Transformers yet. Currently, the only way to run Transformers with sparsity is via demoBERT without using the ONNX parser.

**Comment by ttyio** [comment]
Closing since no activity for more than 3 weeks, please reopen if you still have question, thanks!
